packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -0,0 +1,1432 @@
+/// a [SelectableRegion], consider wrapping that group of widgets using
+/// [SelectionRegistrarScope.disabled].
+///
+/// {@tool dartpad}
+/// This sample demonstrates how to disable selection for a Text in a Column.
+///
+/// ** See code in examples/api/lib/material/selection_area/disable_partial_selection.dart **
+/// {@end-tool}
+///
+/// To create a separate selection system from its parent selection area,
+/// wrap part of the subtree with another [SelectableRegion]. The selection of the
+/// child selection area can not extend pass its subtree, and the selection of
+/// the parent selection area can not extend inside the child selection area.
+///
+/// See also:
+///  * [SelectionArea], which wires up user gestures to enable user selections.
+///  * [SelectionHandler], which contains APIs to handle selection events from the
+///    [SelectableRegion].
+///  * [Selectable], which provides API to participate in the selection system.
+///  * [SelectionRegistrar], which [Selectable] needs to subscribe to receive
+///    selection events.
+///  * [SelectionContainer], which collects selectable widgets in the subtree
+///    and provides api to dispatch selection event to the collected widget.
+class SelectableRegion extends StatefulWidget {
+  /// Create a new [SelectableRegion] widget.
+  ///
+  /// The [selectionControls] is used for building the selection handles and
+  /// toolbar for mobile devices.
+  const SelectableRegion({
+    super.key,
+    required this.focusNode,
+    required this.selectionControls,
+    required this.child,
+  });
+
+  /// {@macro flutter.widgets.Focus.focusNode}
+  final FocusNode focusNode;
+
+  /// The child widget this selection area applies to.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// The delegate to build the selection handles and toolbar for mobile
+  /// devices.
+  final TextSelectionControls selectionControls;
+
+  @override
+  State<StatefulWidget> createState() => SelectableRegionState();
+}
+
+/// State for a [SelectableRegion].
+///
+/// The state provides APIs to modify the selection in a [SelectableRegion].
+class SelectableRegionState extends State<SelectableRegion> with TextSelectionDelegate implements SelectionRegistrar {
+  SelectionOverlay? _selectionOverlay;
+  final LayerLink _startHandleLayerLink = LayerLink();
+  final LayerLink _endHandleLayerLink = LayerLink();
+  final LayerLink _toolbarLayerLink = LayerLink();
+  late _SelectableRegionContainerDelegate _selectionDelegate;
+  // there should only ever be one selectable, which is the SelectionContainer.
+  Selectable? _selectable;
+
+  bool get _hasSelectionOverlayGeometry => _selectionDelegate.value.startSelectionPoint != null
+                                        || _selectionDelegate.value.endSelectionPoint != null;
+
+  @override
+  void initState() {
+    super.initState();
+    _selectionDelegate = _SelectableRegionContainerDelegate();
```

Comment:
nit: could just be instantiated in line 177 and then the late could be replaced with final?